pkg/sql/sqlbase/structured.go

```diff
@@ -3129,6 +3262,97 @@ func (desc *MutableTableDescriptor) MakeMutationComplete(m DescriptorMutation) e
+	oldCol, err := desc.FindColumnByID(swap.OldColumnId)
+	if err != nil {
+		return err
+	}
+	newCol, err := desc.FindColumnByID(swap.NewColumnId)
+	if err != nil {
+		return err
+	}
+
+	// Mark newCol as no longer a computed column.
+	newCol.ComputeExpr = nil
+
+	// oldCol still needs to have values written to it in case nodes read it from
+	// it with a TableDescriptor verison from before the swap.
+	// To achieve this, we make oldCol a computed column of newCol.
+	oldColComputeExpr := tree.CastExpr{
+		Expr:       &tree.ColumnItem{ColumnName: tree.Name(oldCol.Name)},
+		Type:       oldCol.DatumType(),
+		SyntaxMode: tree.CastShort,
+	}
+	s := tree.Serialize(&oldColComputeExpr)
+	oldCol.ComputeExpr = &s
+
+	//Generate unique name for old column.
+	nameExists := func(name string) bool {
+		_, _, err := desc.FindColumnByName(tree.Name(name))
+		return err == nil
+	}
+
+	uniqueName := GenerateUniqueConstraintName(newCol.Name, nameExists)
+
+	// Remember the name of oldCol, because newCol will take it.
+	oldColName := oldCol.Name
+
+	// Rename old column to this new name, and rename newCol to oldCol's name.
+	desc.RenameColumnDescriptor(oldCol, uniqueName)
+	desc.RenameColumnDescriptor(newCol, oldColName)
+
+	// Swap Column Family references.
+	// This preserves the ordering of column families when querying for it.
+	// Since both columns are in the same family, this only affects the
+	// ordering of ColumnIDs in the family.
+	// The old column will be removed from column families
+	// when the column is dropped.
+	for i := range desc.Families {
+		for j := range desc.Families[i].ColumnIDs {
+			if desc.Families[i].ColumnIDs[j] == oldCol.ID {
+				desc.Families[i].ColumnIDs[j] = newCol.ID
+				desc.Families[i].ColumnNames[j] = newCol.Name
+			} else if desc.Families[i].ColumnIDs[j] == newCol.ID {
+				desc.Families[i].ColumnIDs[j] = oldCol.ID
+				desc.Families[i].ColumnNames[j] = oldCol.Name
+			}
+		}
+	}
+
+	// Set newCol's LogicalColumnID to oldCol's LogicalColumnID. This makes
+	// newCol display like oldCol in catalog tables.
+	newCol.LogicalColumnID = oldCol.LogicalColumnID
+	oldCol.LogicalColumnID = 0
+
+	// Mark oldCol as being the result of an AlterColumnType. This allows us
+	// to generate better errors for failing inserts.
+	oldCol.AlterColumnTypeInProgress = true
+
+	// Remove ownership for sequences for the old column otherwise
+	// the owned sequences will be dropped when the old column is dropped.
+	oldCol.OwnsSequenceIds = nil
+
+	// Clone oldColDesc so that we can queue it up as a mutation.
+	// Use oldColCopy to queue mutation in case oldCol's memory address
+	// gets overwritten during mutation.
+	oldColCopy := protoutil.Clone(oldCol).(*ColumnDescriptor)
+	desc.AddColumnMutation(oldColCopy, DescriptorMutation_DROP)
+
+	// Now move newCol into oldCol's position in the descriptor slice
+	// and remove the original entry for newCol in the descriptor slice.
+	for i := range desc.Columns {
```

Comment:
I remember discussing that we might want to first remove new column from the list of columns at the beginning of the method for clarity